QAChat/Data_Processing/confluence_preprocessor.py

```diff
@@ -225,15 +215,62 @@ def add_content_of_pdf_to_all_page_information(self, page_id) -> str:
 
     def load_preprocessed_data(
-        self, before: datetime, after: datetime
+            self, before: datetime, after: datetime
     ) -> List[DataInformation]:
+        self.init_lookup_tables()
+        self.init_blacklist()
         self.get_all_spaces()
         self.get_all_page_ids_from_spaces()
         self.get_relevant_data_from_pages()
-        self.delete_old_content()  # TODO: update
-
+        self.filter_pages()
         return [data for data in self.all_page_information]
 
+    def init_lookup_tables(self):
+        # get the metadata of type Confluence from DB
+        data = self.supabase_client \
+            .table("data_embedding") \
+            .select("metadata") \
+            .eq("metadata->>type", "confluence").execute().data
+
+        for i in data:
+            page_id = i["metadata"]["id"].split("_")[0]
+            chunk_id = i["metadata"]["id"].split("_")[1]
+            last_update = i["metadata"]["last_changed"]
+
+            # add each ID in dict last_update_lookup
+            if page_id not in self.last_update_lookup:
+                self.last_update_lookup[page_id] = datetime.strptime(last_update.split("T")[0], "%Y-%m-%d")
+
+            # add max of chunk ID to chunk_id_lookup_table
+            if page_id not in self.chunk_id_lookup_table:
+                self.chunk_id_lookup_table[page_id] = chunk_id
+            else:
+                if chunk_id > self.chunk_id_lookup_table[page_id]:
+                    self.chunk_id_lookup_table[page_id] = chunk_id
+
+    def filter_pages(self):
+        to_delete = []
+        for i in self.all_page_information:
+            if i.id in self.last_update_lookup:  # if page is already in DB
+                if i.last_changed > self.last_update_lookup[i.id]:  # if there is a change in the page
+                    self.remove_from_db(i.id)  # remove from DB
+                    self.last_update_lookup[i.id] = None  # make the dict's entry None -> To detect remove page
+                elif i.last_changed == self.last_update_lookup[i.id]:  # if no change in the page
+                    to_delete.append(i)  # append in the list
+                    self.last_update_lookup[i.id] = None  # make the dict's entry None -> To detect remove page
+
+        for i in to_delete:
+            self.all_page_information.remove(i)  # remove the page where no changes from the internal list
+
+        for i in self.last_update_lookup:
+            if self.last_update_lookup[i] is not None:  # check which entry is not None -> Page is deleted from website
+                self.remove_from_db(i)  # remove it from DB
+
+    def remove_from_db(self, id):
+        # loop over max number in chunk id and remove all the rows from DB
+        for i in range(0, int(self.chunk_id_lookup_table[id]) + 1):
+            self.supabase_client.table("data_embedding").delete().eq("metadata->>id", str(id) + "_" + str(i)).execute()
+
 
 if __name__ == "__main__":
     cp = ConfluencePreprocessor()

```

Comment:
Do I have to do something before running the main? Like adding a URL somewhere? When I run the main I get this error:
Traceback (most recent call last):
  File "/home/zitrone/Desktop/Amos/amos2023ss03-qachat/QAChat/Data_Processing/confluence_preprocessor.py", line 276, in <module>
    cp = ConfluencePreprocessor()
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/zitrone/Desktop/Amos/amos2023ss03-qachat/QAChat/Data_Processing/confluence_preprocessor.py", line 35, in __init__
    self.confluence = Confluence(
                      ^^^^^^^^^^^
  File "/home/zitrone/Desktop/Amos/amos2023ss03-qachat/venv/lib/python3.11/site-packages/atlassian/confluence.py", line 30, in __init__
    if ("atlassian.net" in url or "jira.com" in url) and ("/wiki" not in url):
        ^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'NoneType' is not iterable

Process finished with exit code 1

Reply:
please use the new version tokens.env. Seems like the url you use URL without wiki